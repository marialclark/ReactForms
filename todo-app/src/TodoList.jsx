import { useState } from 'react';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';
import './styles/TodoList.css';

const TodoList = () => {
  const [todos, setTodos] = useState([]);

  const addTodo = (newTodo) => {
    setTodos((todos) => [...todos, newTodo]);
  };

  const removeTodo = (id) => {
    setTodos((todos) => todos.filter((todo) => todo.id !== id));
  };

  return (
    <div className="TodoList">
      <NewTodoForm addTodo={addTodo} />
      <div className="TodoList-container">
        {todos.map((todo) => (
          <Todo key={todo.id} id={todo.id} task={todo.task} removeTodo={removeTodo} />
        ))}
      </div>
    </div>
  );
};

export default TodoList;