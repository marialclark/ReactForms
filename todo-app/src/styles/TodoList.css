/* Wrapper for the Todo List */
.TodoList {
  margin: 20px auto;
  width: 80%;
  max-width: 500px;
  border: 1px solid #ccc;
  border-radius: 5px;
  padding: 10px;
  background-color: #f9f9f9;
}

.TodoList-container {
  margin-top: 10px;
  display: flex;
  flex-direction: column;
  gap: 5px;
}
