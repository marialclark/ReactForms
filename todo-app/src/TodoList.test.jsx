import { render, screen, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

describe('TodoList Component', () => {
  // Smoke Test
  it('renders without crashing', () => {
    render(<TodoList />);
  });

  // Snapshot Test
  it('matches snapshot', () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
  });

  // Critical Business Logic: Adding a Todo
  it('can add a new todo', () => {
    render(<TodoList />);

    const input = screen.getByLabelText('Task:');
    const button = screen.getByText('Add Todo');
    fireEvent.change(input, { target: { value: 'Test Todo' } });
    fireEvent.click(button);

    expect(screen.getByText('Test Todo')).toBeInTheDocument();
  });

  // Critical Business Logic: Removing a Todo
  it('can remove a todo', () => {
    render(<TodoList />);

    const input = screen.getByLabelText('Task:');
    const button = screen.getByText('Add Todo');
    fireEvent.change(input, { target: { value: 'Test Todo' } });
    fireEvent.click(button);

    const removeButton = screen.getByText('X');
    fireEvent.click(removeButton);

    expect(screen.queryByText('Test Todo')).not.toBeInTheDocument();
  });
});
