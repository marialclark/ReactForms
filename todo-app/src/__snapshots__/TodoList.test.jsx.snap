// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <div
      class="NewTodoForm"
    >
      <form>
        <label
          for="task"
        >
          Task:
        </label>
        <input
          id="task"
          name="task"
          type="text"
          value=""
        />
        <button>
          Add Todo
        </button>
      </form>
    </div>
    <div
      class="TodoList-container"
    />
  </div>
</DocumentFragment>
`;
